File: wrappers/dot_net/pinvoke/OzwOnOffDemo/Log.cs
#region Header
//-----------------------------------------------------------------------------
//
//      Log.cs
//
// 		This class provides logging capabilites to a file and optionally
// 		to the console.  The amount of logging information can be controlled.
//
//      Copyright (c) 2010 Doug Brown <djbrown2001@gmail.com>
//
//      SOFTWARE NOTICE AND LICENSE
//
//      This file is part of OpenZWave.
//
//      OpenZWave is free software: you can redistribute it and/or modify
//      it under the terms of the GNU Lesser General Public License as published
//      by the Free Software Foundation, either version 3 of the License,
//      or (at your option) any later version.
//
//      OpenZWave is distributed in the hope that it will be useful,
//      but WITHOUT ANY WARRANTY; without even the implied warranty of
//      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//      GNU Lesser General Public License for more details.
//
//      You should have received a copy of the GNU Lesser General Public License
//      along with OpenZWave.  If not, see <http://www.gnu.org/licenses/>.
//
//-----------------------------------------------------------------------------
#endregion Header
using System;
using System.IO;

namespace SynapsityAZ.Utils
{
	/// <summary>
	/// This class provides logging capabilites to a file and optionally
	/// to the console.  The amount of logging information can be controlled
	/// by the Level which is one of the Log.Lvl enumerated types which starts
	/// withe Errors at the least verbose level up to Debug which is the most
	/// verbose level.  The methods to create and write the log are static so
	/// instantiation of the class is not necessary.
	/// </summary>
	public class Log
	{
		public enum Lvl
		{
			Errors = 5,				// Least verbose, errors only
			Warnings = 10,
			Minimal = 15,
			Normal = 20,
			Verbose = 30,			// Everything is logged.
			Debug = 40				// Use for debugging
		}
		private static readonly Log instance = new Log();
		private static TextWriter m_tw = null;
		private static Lvl m_logLvl = Lvl.Minimal;						// Contains the current log level
		private static bool m_console = false;

		// Explicit static constructor to tell C# compiler
		// not to mark type as beforefieldinit
		static Log() {}
		
		public static Lvl Level
		{
			set
			{
				m_logLvl = value;
			}
		}
		
		public static bool OutputToConsole
		{
			set
			{
			  	m_console = value;
			}
		}
		
		public static Log Create(String fileName) 
		{
			return Create(fileName, Lvl.Minimal);
		}
		
		public static Log Create(String fileName, Lvl logLvl)
		{
			if (m_tw == null)
			{
				try
				{
					Open(fileName);
					m_logLvl = logLvl;
				}
				catch (Exception e) {
					System.Console.WriteLine("Error opening log file: " + e.ToString());
				}				
			}
			return instance;
		}
		
		private Log()
		{
		}
		
		private static void Open(String fileName)
		{
			m_tw = new StreamWriter(fileName);
		}
		
		public static void Write(String str)
		{
			Write(Lvl.Verbose, str);
		}
		
		public static void Write(Lvl logLvl, String str)
		{
			if (logLvl <= m_logLvl)
			{
				String outStr = DateTime.Now.ToString("MM/dd/yy-HH:mm:ss.fff") + ": " + str;
				m_tw.WriteLine(outStr);
				m_tw.Flush();
				if (m_console)
				{
					System.Console.WriteLine(outStr);
				}
			}
		}
	}
}

File: trunk/cpp/src/platform/Log.cpp
//-----------------------------------------------------------------------------
//
//	Log.cpp
//
//	Cross-platform message and error logging
//
//	Copyright (c) 2010 Mal Lansell <openzwave@lansell.org>
//
//	SOFTWARE NOTICE AND LICENSE
//
//	This file is part of OpenZWave.
//
//	OpenZWave is free software: you can redistribute it and/or modify
//	it under the terms of the GNU Lesser General Public License as published
//	by the Free Software Foundation, either version 3 of the License,
//	or (at your option) any later version.
//
//	OpenZWave is distributed in the hope that it will be useful,
//	but WITHOUT ANY WARRANTY; without even the implied warranty of
//	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//	GNU Lesser General Public License for more details.
//
//	You should have received a copy of the GNU Lesser General Public License
//	along with OpenZWave.  If not, see <http://www.gnu.org/licenses/>.
//
//-----------------------------------------------------------------------------

#include <stdarg.h>

#include "Defs.h"
#include "Mutex.h"
#include "Log.h"

#include "LogImpl.h"	// Platform-specific implementation of a log


using namespace OpenZWave;

Log* Log::s_instance = NULL;
static bool s_dologging;

//-----------------------------------------------------------------------------
//	<Log::Create>
//	Static creation of the singleton
//-----------------------------------------------------------------------------
Log* Log::Create
(
	string const& _filename 
)
{
	if( NULL == s_instance )
	{
		s_instance = new Log( _filename );
		s_dologging = true; // default logging to true so no change to what people experience now
	}

	return s_instance;
}

//-----------------------------------------------------------------------------
//	<Log::Destroy>
//	Static method to destroy the logging singleton.
//-----------------------------------------------------------------------------
void Log::Destroy
(
)
{
	delete s_instance;
	s_instance = NULL;
}

//-----------------------------------------------------------------------------
//	<Log::SetLogggingState>
//	Set flag to actually write to log or skip it
//-----------------------------------------------------------------------------
void Log::SetLoggingState
(
	bool _dologging
)
{
	s_dologging = _dologging;
}

//-----------------------------------------------------------------------------
//	<Log::GetLoggingState>
//	Return a flag to indicate whether logging is enabled
//-----------------------------------------------------------------------------
bool Log::GetLoggingState
(
)
{
	return s_dologging;
}

//-----------------------------------------------------------------------------
//	<Log::Write>
//	Write to the log
//-----------------------------------------------------------------------------
void Log::Write
( 
	char const* _format,
	... 
)
{
	if( s_instance && s_dologging )
	{
	  	s_instance->m_logMutex->Lock();
		va_list args;
		va_start( args, _format );
		s_instance->m_pImpl->Write( _format, args );
		va_end( args );
		s_instance->m_logMutex->Release();
	}
}

//-----------------------------------------------------------------------------
//	<Log::Log>
//	Constructor
//-----------------------------------------------------------------------------
Log::Log
(
	string const& _filename 
):
	m_pImpl( new LogImpl( _filename ) ),
	m_logMutex( new Mutex() )
{
}

//-----------------------------------------------------------------------------
//	<Log::~Log>
//	Destructor
//-----------------------------------------------------------------------------
Log::~Log
(
)
{
	delete m_logMutex;
	delete m_pImpl;
}
